lib/auth/grpcserver.go

```diff
@@ -1065,6 +1066,271 @@ func (g *GRPCServer) DeleteAllKubeServices(ctx context.Context, req *proto.Delet
+			return trace.Wrap(err)
+		}
+	case *proto.MFARegisterResponse_U2F:
+		// u2f.RegisterVerify will upsert the new device internally.
+		dev, err = u2f.RegisterVerify(ctx, u2f.RegisterVerifyParams{
+			DevName: initReq.DeviceName,
+			Resp: u2f.RegisterChallengeResponse{
+				RegistrationData: resp.U2F.RegistrationData,
+				ClientData:       resp.U2F.ClientData,
+			},
+			ChallengeStorageKey:    user,
+			RegistrationStorageKey: user,
+			Storage:                auth.Identity,
+			Clock:                  auth.clock,
+		})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	default:
+		return trail.ToGRPC(trace.BadParameter("MFARegisterResponse sent an unknown response type %v", regResp.Response))
+	}
+
+	// 6. send Ack
+	if err := stream.Send(&proto.AddMFADeviceResponse{
+		Response: &proto.AddMFADeviceResponse_Ack{Ack: &proto.AddMFADeviceResponseAck{Device: dev}},
+	}); err != nil {
+		return trail.ToGRPC(err)
+	}
+	return nil
+}
+
+func (g *GRPCServer) DeleteMFADevice(stream proto.AuthService_DeleteMFADeviceServer) error {
+	ctx := stream.Context()
+	actx, err := g.authenticate(ctx)
+	if err != nil {
+		return trail.ToGRPC(err)
+	}
+	auth := actx.authServer
+	user := actx.User.GetName()
+
+	// The RPC is streaming both ways and the message sequence is:
+	// (-> means client-to-server, <- means server-to-client)
+	//
+	// 1. -> Init
+	// 2. <- MFAChallenge
+	// 3. -> MFAResponse
+	// 4. <- Ack
+
+	// 1. receive client Init
+	req, err := stream.Recv()
+	if err != nil {
+		return trail.ToGRPC(err)
+	}
+	initReq := req.GetInit()
+	if initReq == nil {
+		return trail.ToGRPC(trace.BadParameter("expected DeleteMFADeviceRequestInit, got %T", req))
+	}
+
+	// 2. send MFAAuthenticateChallenge
+	authChallenge, err := auth.mfaAuthChallenge(ctx, user)
+	if err != nil {
+		return trail.ToGRPC(err)
+	}
+	if err := stream.Send(&proto.DeleteMFADeviceResponse{
+		Response: &proto.DeleteMFADeviceResponse_MFAChallenge{MFAChallenge: authChallenge},
+	}); err != nil {
+		return trail.ToGRPC(err)
+	}
+
+	// 3. receive client MFAAuthenticateResponse
+	req, err = stream.Recv()
+	if err != nil {
+		return trail.ToGRPC(err)
+	}
+	authResp := req.GetMFAResponse()
+	if initReq == nil {
+		return trail.ToGRPC(trace.BadParameter("expected MFAAuthenticateResponse, got %T", req))
+	}
+	if err := auth.validateMFAAuthResponse(ctx, user, authResp); err != nil {
+		return trail.ToGRPC(err)
+	}
+
+	// Find the device and delete it from backend.
+	devs, err := auth.GetMFADevices(ctx, user)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	var found bool
+	for _, d := range devs {
+		// Match device by name or ID.
+		if d.Metadata.Name == initReq.DeviceName || d.Id == initReq.DeviceName {
+			found = true
+			if err := auth.DeleteMFADevice(ctx, user, d.Id); err != nil {
+				return trail.ToGRPC(err)
+			}
+			break
+		}
+	}
+	if !found {
+		return trail.ToGRPC(trace.NotFound("MFA device %q does not exist", initReq.DeviceName))
+	}
+
+	// 6. send Ack
```

Comment:
nit: I'd pull the Ack step into the loop and left the error here to clean it up a bit:
```go
for _, d := range devs {
	// Match device by name or ID.
	if d.Metadata.Name == initReq.DeviceName || d.Id == initReq.DeviceName {
		if err := auth.DeleteMFADevice(ctx, user, d.Id); err != nil {
			return trail.ToGRPC(err)
		}
		// 6. send Ack
		if err := stream.Send(&proto.DeleteMFADeviceResponse{
			Response: &proto.DeleteMFADeviceResponse_Ack{Ack: &proto.DeleteMFADeviceResponseAck{}},
		}); err != nil {
			return trail.ToGRPC(err)
		}
		return nil
	}
}
return trail.ToGRPC(trace.NotFound("MFA device %q does not exist", initReq.DeviceName))
```
it could be even more concise with dedicated methods for each step.